src/main/java/com/github/minecraftschurlimods/arsmagicalegacy/client/gui/occulus/OcculusAffinityTabRenderer.java

```diff
@@ -78,22 +91,37 @@ protected void renderFg(PoseStack pMatrixStack, int pMouseX, int pMouseY, float
             getItemRenderer().renderAndDecorateFakeItem(helper.getEssenceForAffinity(aff), drawX + posX, drawY + posY);
             if (pMouseX > drawX && pMouseX < drawX + 16 && pMouseY > drawY && pMouseY < drawY + 16) {
                 drawString.add(aff.getDisplayName().copy().withStyle(style -> style.withColor(aff.getColor())));
-/*
-                    List<AbstractAffinityAbility> abilites = Lists.newArrayList(ArsMagicaAPI.getAffinityAbilityRegistry().getValues());
-                    abilites.sort(new Comparator<AbstractAffinityAbility>() {
-                        @Override
-                        public int compare(AbstractAffinityAbility o1, AbstractAffinityAbility o2) {
-                            return (int) ((o1.getMinimumDepth() * 100) - (o2.getMinimumDepth() * 100));
-                        }
-                    });
-                    for (AbstractAffinityAbility ability : abilites) {
-                        if (ability.getAffinity() == aff) {
-                            String advancedTooltip = "";
-                            if (isShiftDown) advancedTooltip = " (Min. : " + Math.round(ability.getMinimumDepth() * 100) + "%" + (ability.hasMax() ?(", Max. : " + Math.round(ability.getMaximumDepth() * 100) + "%")  : "") + ")";
-                            drawString.add(TextFormatting.RESET.toString() + (ability.isEligible(player) ? TextFormatting.GREEN.toString() : TextFormatting.DARK_RED.toString()) + I18n.translateToLocal("affinityability." + ability.getRegistryName().toString().replaceAll("arsmagica2:", "") + ".name") + advancedTooltip);
+                var abilityManager = api.getAbilityManager();
+                List<ResourceLocation> abilities = abilityManager.getAbilitiesForAffinity(aff);
+                IForgeRegistry<IAbility> abilityRegistry = api.getAbilityRegistry();
+                abilities.sort((o1, o2) -> (int)((abilityManager.get(o1).range().min() * 100) - (abilityManager.get(o2).range().min() * 100)));
+                abilities.forEach(resourceLocation -> {
+                    IAbilityData abilityData = abilityManager.get(resourceLocation);
+                    boolean test = abilityData.test(player);
+                    IAbility value = abilityRegistry.getValue(resourceLocation);
+                    assert value != null;
```

Comment:
Please do not use asserts, return instead.

Reply:
the assert is only there to shush the possible NPE warning it would not throw in prod anyway (at this point it can't be null but getValue is nullable so it cries)